cpp/src/arrow/ipc/ipc-metadata-test.cc

```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include "arrow/ipc/metadata.h"
+#include "arrow/schema.h"
+#include "arrow/test-util.h"
+#include "arrow/type.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+
+class Buffer;
+
+class TestSchemaMessage : public ::testing::Test {
+ public:
+  void SetUp() {}
+
+  void CheckRoundtrip(const Schema* schema) {
+    std::shared_ptr<Buffer> buffer;
+    ASSERT_OK(ipc::WriteSchema(schema, &buffer));
+
+    std::shared_ptr<ipc::Message> message;
+    ASSERT_OK(ipc::Message::Open(buffer, &message));
+
+    ASSERT_EQ(ipc::Message::SCHEMA, message->type());
+
+    std::shared_ptr<ipc::SchemaMessage> schema_msg = message->GetSchema();
+    ASSERT_EQ(schema->num_fields(), schema_msg->num_fields());
+
+    std::shared_ptr<Schema> schema2;
+    ASSERT_OK(schema_msg->GetSchema(&schema2));
+
+    if (!schema2->Equals(*schema)) {
```

Comment:
`operator==` isn't implemented for these types (yet). I'll factor this out into a reusable function